### PR TITLE
Added logging for timeout

### PR DESCRIPTION
**Pull request by anabouna**

### CODE REVIEW

File: src/yaral.ts [hunk 1]
```diff
@@ -278,19 +290,23 @@ export const register: PluginFunction<IYaralOptions> = (
 
     return (err: Error, data?: any) => {
       clearTimeout(timeout);
-      if (timeout !== null) {
+      let isTimedout = timeout === null;
+      //Only log non timed-out requests if timeout is enabled. And even when timed out, still log when call actually resolves
+      server.log(['ratelimit', 'timeout'], getRequestLogDetails(err, req, isTimedout, new Date().getTime() - startTime));
```
Comment (SimonSchick) [inline src/yaral.ts]: Use Date.now() instead of `new Date().getTime()`

File: src/yaral.ts [hunk 2]
```diff
@@ -278,19 +290,23 @@ export const register: PluginFunction<IYaralOptions> = (
 
     return (err: Error, data?: any) => {
       clearTimeout(timeout);
-      if (timeout !== null) {
+      let isTimedout = timeout === null;
```
Comment (SimonSchick) [inline src/yaral.ts]: Shouldn't this be const?

File: src/yaral.ts (hunk 1 above)
Comment (anabouna) [reply src/yaral.ts]: Done. Does ```Date.now()``` have a better performance ?

File: src/yaral.ts (hunk 2 above)
Comment (anabouna) [reply src/yaral.ts]: done

File: src/yaral.ts (hunk 1 above)
Comment (SimonSchick) [reply src/yaral.ts]: It's a more direct call, it also saves the date instantiation.

File: src/yaral.ts [hunk 3]
```diff
@@ -290,9 +290,9 @@ export const register: PluginFunction<IYaralOptions> = (
 
     return (err: Error, data?: any) => {
       clearTimeout(timeout);
-      let isTimedout = timeout === null;
+      const isTimedout = timeout === null;
```
Comment (SimonSchick) [inline src/yaral.ts]: Actually TS should've warned about this, or TSLint, weird.

File: src/yaral.ts (hunk 1 above)
Comment (connor4312) [reply src/yaral.ts]: More idiomatic [and also yes](https://jsperf.com/date-now-vs-new-date)

Comment (anabouna) [reply src/yaral.ts]: much faster ^^  Thanks guys

**Comment by SimonSchick** [comment]
Can you please add a TSLint setup, would avoid some of the problems here.

File: src/yaral.ts (hunk 3 above)
Comment (anabouna) [reply src/yaral.ts]: already installed... weird

File: src/yaral.ts [hunk 4]
```diff
@@ -262,10 +262,22 @@ export const register: PluginFunction<IYaralOptions> = (
...
-
-  const createTimeout = (callback: ((err: Error, data: any) => void)) => {
+  
+  const getRequestLogDetails = (err: Error, req: Request, isTimedout: boolean, duration: number) => {
+    return {
+      url: req.url.href || '',
+      duration: duration,
+      success: !err,
+      properties: {
+        error: err? err.toString() : '',
```
Comment (connor4312) [inline src/yaral.ts]: Prefer `err.stack || err.message`. `.toString()` will not include the stack by default
.

File: src/yaral.ts [hunk 5]
```diff
@@ -377,9 +395,10 @@ export const register: PluginFunction<IYaralOptions> = (
       return reply.continue();
     }
 
-    return limitus.drop(opts.bucket.name(), opts.id, createTimeout((err: Error, data?: any) => {
-      preResponseResolve(err, data, reply, opts, res);
-    }));
+    const startTime = Date.now();
+    return limitus.drop(opts.bucket.name(), opts.id, createTimeout(req, (err: Error, data?: any) => {
+      preResponseResolve(err, data, req, reply, opts, res);
+    }, startTime));
```
Comment (connor4312) [inline src/yaral.ts]: You can actually have createTimeout handle the startTime stuff internally, no need to have it be external. Start the timer when it's created, stop it when the callback fires.

File: src/yaral.ts (hunk 4 above)
Comment (anabouna) [reply src/yaral.ts]: I thought that I just want to know what kind of error it is here (ratelimiting, limitus...) not much interest in the stack. Thoughts ?

Comment (connor4312) [reply src/yaral.ts]: Never hurts to expose extra data. If an error occurs within the handler itself it can be useful to consumers.

Comment (anabouna) [reply src/yaral.ts]: agreed

File: src/yaral.ts (hunk 5 above)
Comment (anabouna) [reply src/yaral.ts]: done

**Pull request merged**